### PR TITLE
feat: replace launch configuration with launch template

### PR DESCRIPTION
**Pull request by Michenux**
## Description

AWS recommend that we should use launch templates instead of launch configurations.

With the launch template, we can now use a CMK for storage encryption. Globally, launch template allows
to configure more things than the launch configuration.

## Migrations required

NO

### CODE REVIEW

**Comment by npalm** [comment]
thanks, will check the PR in the next days, sorry for the delay

**Review by npalm — CHANGES_REQUESTED** [review]
Need a bit more time to check the PR. I ran a quick test with the default example. The plan is failing for this example. In this example `runner_instance_spot_price` is not set. In this case the agent will run on a NON spot instance. Which is the default. This will causse the following error:


```
Error: Invalid function argument

  on ../../main.tf line 219, in resource "aws_launch_template" "gitlab_runner_instance":
 219:     for_each = length(var.runner_instance_spot_price) == 0 ? [] : ["spot"]
 
 ```
 
 Please can you also rebase with develop. I have fixed the CI jobs

**Comment by Michenux** [comment]
I have fixed the issues. Note that i have added an instance_refresh parameter on the runner autoscaling group to recreate the instance runner when a config is changed.

**Comment by npalm** [comment]
sorry for the late response but will check the PR

**Comment by npalm** [comment]
@Michenux great work, sorry for the late response. But I still got the following error while running the default example, see also review comment:

```bash

Error: Invalid function argument

  on ../../main.tf line 233, in resource "aws_launch_template" "gitlab_runner_instance":
 233:     for_each = var.runner_instance_spot_price != null && length(var.runner_instance_spot_price) == 0 ? [] : ["spot"]
    |----------------
    | var.runner_instance_spot_price is null

Invalid value for "value" parameter: argument must not be null.


Error: Invalid function argument

  on ../../main.tf line 271, in resource "aws_launch_template" "gitlab_runner_instance":
 271:     for_each = var.runner_instance_spot_price != null && length(var.runner_instance_spot_price) == 0 ? [] : ["spot"]
    |----------------
    | var.runner_instance_spot_price is null

Invalid value for "value" parameter: argument must not be null.
```

Did you fixed this issue? See only commits about formatting after my review.

File: main.tf [hunk 1]
```diff
@@ -205,30 +218,63 @@ data "aws_ami" "runner" {
...
-  ebs_optimized        = var.runner_instance_ebs_optimized
-  enable_monitoring    = var.runner_instance_enable_monitoring
-  spot_price           = var.runner_instance_spot_price
-  iam_instance_profile = aws_iam_instance_profile.instance.name
-  dynamic "root_block_device" {
+resource "aws_launch_template" "gitlab_runner_instance" {
+  name_prefix            = var.runners_name
+  key_name               = var.ssh_key_pair
+  image_id               = data.aws_ami.runner.id
+  user_data              = local.template_user_data
```
Comment (npalm) [inline main.tf:225]:
```suggestion
  user_data              = base64encode(local.template_user_data)
```

File: main.tf [hunk 2]
```diff
@@ -205,30 +218,63 @@ data "aws_ami" "runner" {
...
+  instance_type          = var.instance_type
+  update_default_version = true
+  ebs_optimized          = var.runner_instance_ebs_optimized
+  monitoring {
+    enabled = var.runner_instance_enable_monitoring
+  }
+  dynamic "instance_market_options" {
+    for_each = var.runner_instance_spot_price != null && length(var.runner_instance_spot_price) == 0 ? [] : ["spot"]
```
Comment (npalm) [inline main.tf:233]:
```suggestion
    for_each = var.runner_instance_spot_price == null ? [] : ["spot"]
```

File: main.tf [hunk 3]
```diff
@@ -205,30 +218,63 @@ data "aws_ami" "runner" {
...
+  tag_specifications {
+    resource_type = "instance"
+    tags          = local.tags
+  }
+  tag_specifications {
+    resource_type = "volume"
+    tags          = local.tags
+  }
+  dynamic "tag_specifications" {
+    for_each = var.runner_instance_spot_price != null && length(var.runner_instance_spot_price) == 0 ? [] : ["spot"]
```
Comment (npalm) [inline main.tf:271]:
```suggestion
    for_each = var.runner_instance_spot_price == null ? [] : ["spot"]
```

**Review by npalm — CHANGES_REQUESTED** [review]
got it working locally, need to test a bit more, sse suggestion for my changes

File: main.tf (hunk 3 above)
Comment (Michenux) [reply main.tf:271]: I would prefer to keep the test with the empty string.

Should work : 
```suggestion
    for_each = var.runner_instance_spot_price == null || var.runner_instance_spot_price == "" ? ["spot"] : []
```

**Comment by npalm** [comment]
Please can you also rebase develop, change in PR #348 requires a small update in the launch template.

**Comment by Michenux** [comment]
I have rebased from develop and applied another solution for var.runner_instance_spot_price condition.

File: main.tf (hunk 2 above)
Comment (npalm) [reply main.tf:233]: I still get an error on this line, aslo the for_each for the tag_specification looks not correct. Shoul the loop not only run onse the agent instance / runner agent instance is configured as spot by setting `runner_instance_spot_price`. The code in the PR leads to a raise condition on the default example. I have changed the for_each line to:

```
for_each = var.runner_instance_spot_price == null || var.runner_instance_spot_price == "" ? [] : ["spot"]
```
After this change and the change to tag_specification it looks working like a charm. Tested for both spot and non spot.

File: main.tf [hunk 4]
```diff
@@ -205,34 +218,69 @@ data "aws_ami" "runner" {
...
+  tag_specifications {
+    resource_type = "instance"
+    tags          = local.tags
+  }
+  tag_specifications {
+    resource_type = "volume"
+    tags          = local.tags
+  }
+  dynamic "tag_specifications" {
+    for_each = var.runner_instance_spot_price == null || var.runner_instance_spot_price == "" ? ["spot"] : []
```
Comment (npalm) [inline main.tf:271]:
```suggestion
    for_each = var.runner_instance_spot_price == null || var.runner_instance_spot_price == "" ? [] : ["spot"]
```

**Review by npalm — CHANGES_REQUESTED** [review]
@Michenux thanks for all the hard work, sorry for the delay (once again). The current code is not working, once update the for_each on 2 places to `for_each = var.runner_instance_spot_price == null || var.runner_instance_spot_price == "" ? [] : ["spot"]` the default example works. Tested for spot and non spot.

File: main.tf [hunk 5]
```diff
@@ -205,34 +218,69 @@ data "aws_ami" "runner" {
...
+  image_id               = data.aws_ami.runner.id
+  user_data              = base64encode(local.template_user_data)
+  instance_type          = var.instance_type
+  update_default_version = true
+  ebs_optimized          = var.runner_instance_ebs_optimized
+  monitoring {
+    enabled = var.runner_instance_enable_monitoring
+  }
+  dynamic "instance_market_options" {
+    for_each = var.runner_instance_spot_price != null && length(var.runner_instance_spot_price) == 0 ? [] : ["spot"]
```
Comment (npalm) [inline main.tf:233]: The array should be set with spot once the runner_instance_spot_price is set, same as for tag_specification
```suggestion
    for_each = var.runner_instance_spot_price == null || var.runner_instance_spot_price == "" ? [] : ["spot"]
```

**Review by npalm — CHANGES_REQUESTED** [review]
Looks like you missed on configuration for the istance market option, have made a suggestion with the fix. 

Did you test one ot the examples?

File: main.tf (hunk 5 above)
Comment (npalm) [reply main.tf:233]: Tested with this changes and it looks all good!

**Review by npalm — APPROVED** [review]

**Pull request merged**